RFC: make Cargo embed dependency versions in the compiled binary (#2801)
rust-lang/rfcs: text/0000-cargo-embed-dependency-versions.md

Linux distributions alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.

The primary use case for this information is cross-referencing versions of the dependencies against [RustSec advisory database](https://github.com/RustSec/advisory-db) and/or third-party databases such as [Common Vulnerabilities and Exposures](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). This also enables use cases such as ensuring a fix in a depencency has been propagated across the entirety of your fleet or preventing binaries with unvetted dependencies from accidentally reaching a production environment - all with zero bookkeeping.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or runtime), and origin (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.

Reviewer: > and origin

nit: this makes it sound like full origin information is encoded.  This is clarified later but you might lose some readers before then.  Clarifying this to be something like "and origin kind" or something could help

Author: Done

This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.

WASM, asm.js and embedded platforms are excempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.

> A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).

Reviewer: I'm reading the RFC as saying this config option exists.  If so, I would expect the relevant `Cargo.toml` or `config.toml` schema to be specified in the RFC

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is subject to bikeshedding.

The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.

> The JSON schema specifying the format is provided below. If you find Rust structures more readable, you can find them [here](https://github.com/rust-secure-code/cargo-auditable/blob/311f9932128667b8b18113becdea276b3d98aace/auditable-serde/src/lib.rs#L99-L172). In case of divergences the JSON schema provided in this RFC takes precedence.

Reviewer: Thank you for that link!  I find reading schemas to not be the easiest way to consume information

Reviewer 2: Here's a [`cargo-binstall`'s audtiable data](https://gist.github.com/NobodyXu/f904d194c670549ee1c94bf759a9353f), hope you find it useful when paired with the schema.

Reviewer 2: Here's a part of it for a rough idea:

```json
  "packages": [
    {
      "name": "adler",
      "version": "1.0.2",
      "source": "crates.io"
    },
    {
      "name": "ahash",
      "version": "0.7.6",
      "source": "crates.io",
      "dependencies": [
        93,
        202,
        323
      ]
    },
    {
      "name": "ahash",
      "version": "0.8.3",
      "source": "crates.io",
      "dependencies": [
        42,
        93,
        202,
        323
      ]
    },
    {
      "name": "aho-corasick",
      "version": "1.0.2",
      "source": "crates.io",
      "dependencies": [
        186
      ]
    },
```

```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "required": [
        "packages"
    ],
    "properties": {
        "packages": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/Package"
            }
        }
    },
    "definitions": {
        "DependencyKind": {
            "type": "string",
            "enum": [
                "build",

>                 "runtime"

Reviewer: Normally I would consider a "runtime" dependency to be dynamically linked.

In cargo, we call these "normal" dependencies which isn't very clear :)

e.g. from `cargo tree`)
```
  -e, --edges <KINDS>         The kinds of dependencies to display (features, normal, build, dev, all, no-normal,
                              no-build, no-dev, no-proc-macro)
```

Author: At this point renaming that causes a format break and requires the whole ecosystem around this (cargo-audit, Syft, Trivy) to transition to these new names. I'm not opposed to this on principle, but we should make sure the changes are worth the breakage.

Reviewer 2: Perhaps you can make `runtime` a deprecated alias of `normal`?

Reviewer 2: That's true, but it breaks less than directly renaming `runtime` => `normal`.

Reviewer 3: Honestly, this is sounding like the "Firefox responds to certain `-webkit-`-prefixed CSS properties" de-facto standardization situation that Rust's system of RFCs and nightlies and `#[feature]` was designed to avoid.

Reviewer 2: > This is still a breaking change. It would require changing all pasers of this data to recognize the `normal` variant.

@Shnatsel Perhaps you can add this to the `auditable` and `audit` implementation to `cargo` nightly only, then add support for it in `cargo-audit`.

Once it stablised, you can then deprecated `cargo-auditable` and `cargo-audit` for latest `cargo` and all the previous `cargo-audit` versions can support decoding embedded information encoded by `cargo`.

Reviewer 2: > Honestly, this is sounding like the "Firefox responds to certain `-webkit-`-prefixed CSS properties" de-facto standardization situation that Rust's system of RFCs and nightlies and `#[feature]` was designed to avoid.

Yeah it's a bit unfortunate.

Author: In practice the "runtime" string never appears in the resulting. It is the default case and it is omitted to save space. So the point is rather moot anyway.

Author: I've changed it from "runtime" to "normal" for consistency with the help text.

                        {
                            "$ref": "#/definitions/Source"
                        }
                    ]
                },
                "version": {
                    "description": "The package's version in the [semantic version](https://semver.org) format.",
                    "type": "string"
                }
            }
        },
        "Source": {
            "description": "Serializes to \"git\", \"local\", \"crates.io\" or \"registry\". May be extended in the future with other revision control systems, etc.",
            "oneOf": [
                {
                    "type": "string",
                    "enum": [
                        "crates.io",
                        "git",
                        "local",
                        "registry"
                    ]
                },
            ]
        }
    }
}
```

# Drawbacks
[drawbacks]: #drawbacks

- Slightly increases the size of the generated binaries. However, the increase is [typically below 1%](https://github.com/rust-lang/rfcs/pull/2801#issuecomment-549184251).
- Adds more platform-specific code to the build process, which needs to be maintained.

> # Rationale and alternatives

Reviewer: What about alternative [SBOM formats](https://github.com/rust-secure-code/cargo-auditable/issues/31)?

Author: They are not designed for this use case, and we found them unsuitable for a variety of reasons - the size too large, they require a timestamp that disrupts reproducible builds, etc.

[Trivy](https://github.com/aquasecurity/trivy) can be used to convert from this internal format to other SBOM formats, right now. I believe external tooling that converts between formats is the way to go. 

The proposed format is similar to other SBOMs in structure, so supporting it in tooling that deals with SBOMs is rather trivial (or at least it was in Trivy).

Author: Added to the RFC

Alternatives:

- Do nothing.
  - Identifying vulnerable binaries will remain impossible. We will see increasing number of known vulnerabilities unpatched in production.
- Track version information separately from the binaries, recording it when running `cargo install` and surfacing it through some other Cargo subcommand. When installing not though `cargo install`, rely on Linux package managers to track version information.
  - Identifying vulnerable binaries will remain impossible on all other platforms, as well as on Linux for code compiled with `cargo build`.
  - Verification by third parties will remain impossible.
- Record version information in a `&'static str` in the binary instead if ELF sections, with start/stop markers to allow black-box extraction from the outside.
  - This has been [prototyped](https://github.com/Shnatsel/rust-audit). It has the upside of allowing the binary itself to introspect its version info with little parsing, but the extraction is less efficient, and this is harder to implement and maintain.

> - Provide a Cargo wrapper or plugin to implement this, but do not put it in Cargo itself.

Reviewer: I think its also worth pointing out the gaps that occur when this isn't first-party, like https://github.com/rust-secure-code/cargo-auditable/issues/124 which happened because `cargo-auditable` is trying to interpret information in a way its not intended to but can't get any other way

Author: Added to the RFC

Go compiler embeds `go.mod` dependency information into its compiled binaries. Due to Go binaries generally being far larger than Rust binaries, the binary size is not a constraint, so they embed much more information - e.g. the licence for each package in the dependency tree, which is then read by the [golicense](https://github.com/mitchellh/golicense) tool.

The most common way to manage Ruby apps involves `Gemfile.lock` which can be thought of as a runtime `Cargo.lock`. Some companies have automation searching for these files in production VMs/containers and cross-referencing them against [RubySec](https://rubysec.com/).

Since build system and package management system are usually decoupled, most other languages did not have the opportunity to implement anything like this.

In microservice environments it is fairly typical to expose an HTTP endpoint returning the application version, see e.g. [example from Go cookbook](https://blog.kowalczyk.info/article/vEja/embedding-build-number-in-go-executable.html). However, this typically does not include versions of the dependencies.

> # Unresolved questions

Reviewer: What about future compat for an unsupported platform that becomes supported?

Author: I imagine it would go through the same staging process - a lengthy period of canarying on nightly followed by a rollout to stable with an announcement in the release notes.

I'm not sure I understand the question, so please let me know if there's something I didn't address.

This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.

WASM, asm.js and embedded platforms are excempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.

A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is subject to bikeshedding.

Reviewer: Can you call out specifically the section name fo we can do that bike shedding or call out in unresolved questions

Author: Done

> The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.

Reviewer: Do we need format versioning? With the entry name or in the schema?

Author: Good question! The section name is versioned, to allow potentially transitioning away from JSON+Zlib in the future. I'll add it to the RFC.

> The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.

Reviewer 4: How much does the zlib compression actually save?

Author: I'm seeing a reduction from 10kb to 2kb on the `cargo audit` dependency information. So, 5x savings. Very much worth it.

Reviewer 2: For `cargo-binstall`, the decompressed info is about 28.9KB, after compressing using `gzip -9` it is only 5KB large, so definitely worth it.

https://github.com/rust-lang/rfcs/pull/2801#issuecomment-1665073360

The JSON schema specifying the format is provided below. If you find Rust structures more readable, you can find them [here](https://github.com/rust-secure-code/cargo-auditable/blob/311f9932128667b8b18113becdea276b3d98aace/auditable-serde/src/lib.rs#L99-L172). In case of divergences the JSON schema provided in this RFC takes precedence.

```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "required": [
        "packages"
    ],
    "properties": {
        "packages": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/Package"
            }
        }
    },
    "definitions": {
        "DependencyKind": {
            "type": "string",
            "enum": [
                "build",
                "runtime"
            ]
        },
        "Package": {
            "description": "A single package in the dependency tree",
            "type": "object",
            "required": [
                "name",
                "source",
                "version"
            ],
            "properties": {

>                 "dependencies": {

Reviewer 4: It might be too verbose, but would it be worth recording the feature flags for each crate (in case there is a vuln associated with a specific feature being enabled)?

Author: I have included that in the initial prototype, but eventually removed it. 

It increased the data size by nearly 2x. Also, nothing was actually consuming this information - neither rustsec.org that powers `cargo audit` nor any other vulnerability database allows expressing the concept that only a specific feature is affected.

However, this can be added later without breaking existing consumers if there are compelling enough use cases to justify the binary size overhead.

Reviewer 5: I'm not advocating for including the feature information if nothing actually needs it, but I have questions on two points.

> It increased the data size by nearly 2x.

Was this with all the feature names included as strings in the JSON? The set of features for a given package (name, version, source)-triple is knowable out-of-band, right (from the source's index)? So the data on which features are enabled for a package could be stored as, e.g., a base64-encoded bitset, with each bit representing whether a feature is enabled, with the bits in the order of the package's sorted feature names?

> neither rustsec.org [...] allows expressing the concept that only a specific feature is affected.

If a RustSec advisory marks only certain functions as affected, and all those functions are behind a feature flag (or set of feature flags), wouldn't that express that only that feature is affected (or that only those features are affected)?

Author: > Was this with all the feature names included as strings in the JSON?

Yes.

> The set of features for a given package (name, version, source)-triple is knowable out-of-band, right (from the source's index)? So the data on which features are enabled for a package could be stored as, e.g., a base64-encoded bitset, with each bit representing whether a feature is enabled, with the bits in the order of the package's sorted feature names?

This would work for packages from crates.io, at the cost of requiring the crates.io index to recover this information. This would not work for packages from a private registry.

> If a RustSec advisory marks only certain functions as affected, and all those functions are behind a feature flag (or set of feature flags), wouldn't that express that only that feature is affected (or that only those features are affected)?

Theoretically yes, but the tooling to recover function-feature relationships does not exist yet.

Author: The list of features can be added later in a backwards-compatible way, if there are compelling use cases for it, since it's just an extra field.

>                 "dependencies": {

Reviewer 4: Another piece of information that may be useful is the flags provided to the compiler (which would allow checking if various mitigations, like control-flow guard and retpoline, were enabled).

Author: This can already be detected by dedicated tools without requiring this information to be explicitly recorded at compile time: https://github.com/slimm609/checksec.sh https://github.com/trailofbits/winchecksec etc.

- Any binary authentication that might be deployed automatically applies to the version information. There is no need to separately authenticate it.
- Tooling for extracting information from binaries (such as ELF sections) is already readily available, as are zlib decompressors and JSON parsers. It can be extracted and parsed [in 5 lines of Python](https://github.com/rust-secure-code/cargo-auditable/blob/master/PARSING.md), or even with a shell one-liner in a pinch.
- This enables third parties such as cloud providers to scan your binaries for you. Google Cloud [already provides such a service](https://cloud.google.com/container-registry/docs/get-image-vulnerabilities), Amazon has [an open-source project you can deploy](https://aws.amazon.com/blogs/publicsector/detect-vulnerabilities-in-the-docker-images-in-your-applications/) while Azure [integrates several partner solutions](https://docs.microsoft.com/en-us/azure/security-center/security-center-vulnerability-assessment-recommendations). They do not support this specific format yet, but integration into Trivy was very easy, so adding support will likely be trivial.

> Alternatives:

Reviewer 4: Could you please add a note about recording the version information in the debug symbols? This has the benefit of avoiding the binary cost (when debug symbols are stripped or separate) but potentially allowing the info to be accessed via whatever mechanism is provided for getting debug symbols (e.g., symbol servers on Windows).

Author: I do not see a compelling reason for storing this in debug symbols. However, I see many drawbacks to that approach./

There are use cases for keeping this data when debug symbols are stripped. For example, many Linux distributions ship debug symbols in separate packages. We want the binaries to contain this data, not the separately shipped debug symbols.

If encoding into debug symbols were to be supported, then there would be two different places where this information could be located. This would require more complex extraction code.

Debug symbols are also highly platform-specific and have many variations even within one platform. To make matters worse, Windows PDB format is undocumented. Parsing debug info would be much more difficult than extracting a binary section, even if that were the *only* place where this information was recorded.

Author: I've added this to the RFC.

  - For compatibility with systems that cannot consume this data directly, external tools can be used to convert to industry standard SBOMs. [Syft](https://github.com/anchore/syft) can already do this today.
- Provide a Cargo wrapper or plugin to implement this, but do not put it in Cargo itself.
  - Third-party implementations cannot be perfectly reliable because Cargo does not expose sufficient information for a perfectly robust system. For example, custom target specifications are impossible to support. There are also [other corner cases](https://github.com/rust-secure-code/cargo-auditable/issues/124) that appear to be impossible to resolve based on the information from `cargo metadata` alone.
  - When people actually need this information (e.g. to check if they're impacted by a vulnerability) it is too late to reach for third-party tooling - the executables have already been built and deployed, and the information is already lost. As such, this mechanism is ineffective if it's not enabled by default.

# Prior art
[prior-art]: #prior-art

> An out-of-tree implementation of this RFC exists, see [`cargo auditable`](https://github.com/rust-secure-code/cargo-auditable/), and has garnered considerable interest. NixOS and Void Linux build all their Rust packages with it today; it is also used in production at Miscrosoft. Extracting the embedded data is already supported by [`rust-audit-info`](https://crates.io/crates/rust-audit-info) and [Syft](https://github.com/anchore/syft). Auditing such binaries for known vulnerabilities is already supported by [`cargo audit`](https://crates.io/crates/cargo-audit) and [Trivy](https://github.com/aquasecurity/trivy).

Reviewer 4: ```suggestion
An out-of-tree implementation of this RFC exists, see [`cargo auditable`](https://github.com/rust-secure-code/cargo-auditable/), and has garnered considerable interest. NixOS and Void Linux build all their Rust packages with it today; it is also used in production at Microsoft. Extracting the embedded data is already supported by [`rust-audit-info`](https://crates.io/crates/rust-audit-info) and [Syft](https://github.com/anchore/syft). Auditing such binaries for known vulnerabilities is already supported by [`cargo audit`](https://crates.io/crates/cargo-audit) and [Trivy](https://github.com/aquasecurity/trivy).
```

This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.

WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.

A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is `.dep-v0` across all platforms. The section name must be changed if breaking changes are made to the format.

Reviewer: Going to re-raise a concern from earlier in the thread

> Given that this uses a completely Rust-specific (and cargo-specific) format, the section name should not use something generic like .dep-list.

(now `.dep-v0`)

Reviewer: Should we separately encode schema + version + compression?

Author: Allowing permutations of schema+version+compression methods is an explicit non-goal. Experiments with other compression methods show that they provide little to no gain over zlib for this use case. JSON is carefully chosen to be easy to consume both programmatically and by humans in a pinch.

Allowing more format variations would also complicate parsers - they would have to support several variants of the format, which leads to more attack surface and more room for bugs.

Author: I've noted in the RFC that the exact name of the section is subject to bikeshedding.

Reviewer 5: > Note that there is a hard limit of 8 bytes on the name

Am I right to think that that still would allow, e.g., `.rsdeps0` (or `.rsDeps0` if upper-case letters work)?  If a version number > 9 is needed, the "s" could be dropped.

Reviewer 5: > Note that there is a hard limit of 8 bytes on the name

Am I right to think that that still would allow, e.g., `.rsdeps0` (or `.rsDeps0` if upper-case letters work)? If a version number > 9 is needed, `deps` could be shortened further to `dep`.

Author: Yes, that is permitted by the section naming constraints.

> The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.

Reviewer: What is the build-time impact of this change?

I know first-party will be faster but still wondering about the overhead.  Both for debug builds generally but also with the upcoming cargo-script

I wonder if we should enable it on a per-profile basis so it can be enabled for release builds but not debug builds

Reviewer 2: [`cargo-binstall` CI](https://github.com/cargo-bins/cargo-binstall/actions/runs/5758584682/job/15611362302) has been using `cargo-auditable` for release build and it seems that impact on build time is insignificant.

Author: You can measure the impact right now with this command:

```bash
hyperfine --max-runs 5 --prepare 'cargo clean' 'cargo build' 'cargo auditable build'
```

It shows a 3% build time regression when building `cargo auditable` itself in debug mode, or a 400ms increase. A large portion of that time is spent in a call to `cargo metadata` and repeated calls to `rustc -vV`. (This is on a system with `rustup`, so most of that time is spent in `rustup` parsing its configuration on every call to `rustc`). 

`cargo auditable` itself is so fast that it doesn't even show up on the profile with a 1000Hz sampling interval: https://share.firefox.dev/4529Yq8

The code is also quite naive right now. I have not attempted to optimize it in any way because it was already fast enough, so there are performance gains to be had if build time ever becomes a problem. The only thing I have tried using lighter Zlib compression in debug mode builds, but this did not result in any measurable changes because the files being compressed are just dozens of kilobytes and that's near-instant regardless.

Author: On a [large workspace](https://github.com/rustsec/rustsec) `cargo auditable` actually comes out 5% faster on my machine, so the compile time differences must be within noise threshold.

Reviewer: Thanks! And to be clear, I expect this to not be too noticeable for most projects but I am protective of build times for "cargo script".

Reviewer: I still wonder if we should distinguish between debug and release builds.

Reviewer 2: > Thanks! And to be clear, I expect this to not be too noticeable for most projects but I am protective of build times for "cargo script".

I think it does not make sense to enable this for "cargo script", given that it is distributed by source code instead of the pre-built artifacts?

Author: I don't think it's possible to decide at the design stage, without trying it and seeing what the impact actually is.

I can add a sentence about it being always enabled for release builds, and for debug builds only if the compilation time impact is deemed to be negligible, how does that sound?

Reviewer: Id rather not hack up cargo with special cases for "cargo script" and we do plan to eventually allow publishing them which means you can `cargo install` one, one day.

With that long term route in mind, we shouldn't preclude "cargo scripts" from using this feature. However, they are run in debug mode by default, so if we make put the setting in the profile and have it be different for debug vs release, then day-to-day "cargo scripts" won't use it.

Author: I have included the build time considerations into the RFC, and added a clause that if the compile time impact ends up being noticeable, this feature will be disabled for debug build by default prior to stabilization.

> The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.

Reviewer 6: Unless there's a specific established standard we need to be compatible with here, can we please use zstd for this? It's typically smaller, typically faster to compress for the same size, and almost always faster to decompress.

Author: This has been tested, and the gains from zstd compression are [negligible](https://github.com/rust-lang/rfcs/pull/2801#issuecomment-1665073360) even at extreme compression ratios.

The decompression speed does not really matter because the data is very small (under 50kb uncompressed, even in sprawling dependency trees). It is a negligible part of the parsing time.

Zstd does have drawbacks, however. For one, zstd does not have a Rust compressor implementation, and its Rust decompressor implementation is slow (3x slower than the reference C impl). This creates a highly non-trivial, memory-unsafe C dependency for all consumers of this format. 

So using zstd will complicate parsing and create far more attack surface for no real gain.

Reviewer 6: FWIW I would love to have a pure-Rust zstd compressor and fast decompressor. I'm not willing to block on this, though. If a pure-Rust version becomes available in the future, perhaps dep-v1 can switch to it.

Reviewer 7: Object uses the ruzstd crate for decompressing zstd compressed sections. Ruzstd is a pure rust zstd decompressor. According to it's readme it is about 1.4-3.5x slower than the reference implementation.

Reviewer 2: @joshtriplett Using zstd instead of zlib doesn't provide much benefit at all:

> I used `rust-audit-info` to extract auditable data embedded in [`cargo-binstall-aarch64-apple-darwin`](https://github.com/cargo-bins/cargo-binstall/releases/download/v1.1.2/cargo-binstall-aarch64-apple-darwin.zip).
> 
> The decompressed data is 28.9KB, after compressed by `gzip -9` it's about 5KB.
> Compressing it using `zstd --ultra -22 --long` takes about 4.8KB, apparently I was wrong about suggestion of using zstd instead of zlib sinc `zstd` produce data set that is only 0.2KB smaller.
> 
> Given that `cargo-binstall` has roughly 318 dependencies, 5KB is reasonable since our pre-built artifacts are already quite large (9.6M after decompressed).

Originally posted at https://github.com/rust-lang/rfcs/pull/2801#issuecomment-1665073360

  - "SPDX in Zlib in a linker section" is not really an industry-standard format. Adding support for the custom format to [Syft](https://github.com/anchore/syft) was trivial, since it's nearly isomorphic to other SBOM formats, so the custom JSON encoding does not seem to add a lot of overhead to consuming this data.
  - For compatibility with systems that cannot consume this data directly, external tools can be used to convert to industry standard SBOMs. [Syft](https://github.com/anchore/syft) can already do this today.
- Record version information in debug symbols instead of binary sections.
  - Debug information formats are highly platform-specific, complex, and poorly documented. For example, Microsoft provides no documentation for Windows PDB. Extracting it would be considerably more difficult. Parsing debug information would be a major source of complexity and bugs.
  - Some Linux distributions, such as Debian, ship debug symbols separately from the binaries, and do not install the debug symbols by default. We need this information to be included in the binaries, not the debug symbols.
- Provide a Cargo wrapper or plugin to implement this, but do not put it in Cargo itself.

>   - Third-party implementations cannot be perfectly reliable because Cargo does not expose sufficient information for a perfectly robust system. For example, custom target specifications are impossible to support. There are also [other corner cases](https://github.com/rust-secure-code/cargo-auditable/issues/124) that appear to be impossible to resolve based on the information from `cargo metadata` alone.

Reviewer: I assume first-party will also be faster because we won't need to do a redundant `cargo metadata` call

Author: Indeed. There are even more savings to be had by avoiding the repeated `rustc -vV` calls, which are costly on systems with `rustup` because `rustup` has to parse its configuration every time to select the right `rustc` version.

Go compiler embeds `go.mod` dependency information into its compiled binaries. Due to Go binaries generally being far larger than Rust binaries, the binary size is not a constraint, so they embed much more information - e.g. the licence for each package in the dependency tree, which is then read by the [golicense](https://github.com/mitchellh/golicense) tool.

The most common way to manage Ruby apps involves `Gemfile.lock` which can be thought of as a runtime `Cargo.lock`. Some companies have automation searching for these files in production VMs/containers and cross-referencing them against [RubySec](https://rubysec.com/).

Since build system and package management system are usually decoupled, most other languages did not have the opportunity to implement anything like this.

In microservice environments it is fairly typical to expose an HTTP endpoint returning the application version, see e.g. [example from Go cookbook](https://blog.kowalczyk.info/article/vEja/embedding-build-number-in-go-executable.html). However, this typically does not include versions of the dependencies.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> - How exactly this should be enabled or disabled for a given target? Should there be a flag in the target configuration file, or some other mechanism?

Author: I simply lack the relevant experience with the target specification files to be able to propose anything specific here.

I would be happy to work this out with someone familiar with the relevant infrastructure prior to merging the RFC.

> - How exactly this should be enabled or disabled for a given target? Should there be a flag in the target configuration file, or some other mechanism?

Author: For example, would the target specification JSON files be a good place to track whether a target supports embedding such metadata into the binaries?

> - How exactly this should be enabled or disabled for a given target? Should there be a flag in the target configuration file, or some other mechanism?

Author: I've specified in the RFC that the target specification JSON controls whether this is done for a given platform or not.

> - How exactly should opt-in or opt-out from embedding this data be toggled? Should it be per-profile, like `strip = true`, or a global configuration option?

Reviewer 6: Per profile, definitely. That will allow people to opt into shipping this information in specific build configurations, while omitting it in others.

Author: Does per-platform configuration make sense? Could we have it configured for *both* certain platforms and certain build profiles?

Reviewer 6: Perhaps, but not more so than other profile settings. In general we would probably benefit from a mechanism to have target-specific profile settings, but I probably don't think we need that as a prerequisite for this.

Author: Thanks! I've updated the RFC to specify this.

Linux distributions alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.

The primary use case for this information is cross-referencing versions of the dependencies against [RustSec advisory database](https://github.com/RustSec/advisory-db) and/or third-party databases such as [Common Vulnerabilities and Exposures](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). This also enables use cases such as ensuring a fix in a depencency has been propagated across the entirety of your fleet or preventing binaries with unvetted dependencies from accidentally reaching a production environment - all with zero bookkeeping.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or normal), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
> 
> This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
> 
> WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
> 
> A per-profile configuration option in `Cargo.toml` can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)). The exact name of this option is subject to bikeshedding.

Reviewer 6: ```suggestion
Every time an executable is compiled with Cargo, Cargo can record the dependency tree of the executable in a special section in the binary. This includes the names, versions, dependency kind (build or normal), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.

This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can then be used to create a Software Bill of Materials (SBOM) in a common format, or audit the dependency list for known vulnerabilities.

A per-profile configuration option in `Cargo.toml` can be used to opt into this behavior when desired (e.g. when not concerned about binary size or information disclosure). The exact name of this option is subject to bikeshedding.

For users who don't enable this option, we encourage using tooling that separately records the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
```

(And perhaps an addition to the future-work section saying that Cargo itself could help with the separate-database case, perhaps by emitting the dep information in a separate file rather than embedded, so that the user can record it.)

Toggling this from opt-out to opt-in.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is `.dep-v0` across all platforms (subject to bikeshedding, but [within 8 bytes](https://github.com/rust-lang/rust/blob/4f7bb9890c0402cd145556ac1929d13d7524959e/compiler/rustc_codegen_ssa/src/back/metadata.rs#L462-L475)). The section name must be changed if breaking changes are made to the format.

> The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.

Reviewer 6: ```suggestion
The data is encoded in JSON which is compressed with Zlib. All arrays are sorted to not disrupt reproducible builds.
```

                    "type": "boolean"
                },
                "source": {
                    "description": "Currently \"git\", \"local\", \"crates.io\" or \"registry\". May be extended in the future with other revision control systems, etc.",
                    "allOf": [
                        {
                            "$ref": "#/definitions/Source"
                        }
                    ]
                },
                "version": {
                    "description": "The package's version in the [semantic version](https://semver.org) format.",
                    "type": "string"
                }
            }
        },
        "Source": {
            "description": "Serializes to \"git\", \"local\", \"crates.io\" or \"registry\". May be extended in the future with other revision control systems, etc.",
            "oneOf": [
                {
                    "type": "string",
                    "enum": [
                        "crates.io",
                        "git",
                        "local",
                        "registry"
                    ]
                },
            ]
        }
    }
}
```

> Not all compilations targets will embed this data. Some may support it but disable it by default (e.g. WebAssembly) while others may not support it at all. Whether the target support it, and whether embedding this data is enabled by default for a given target is recorded in the [target specification JSON](https://doc.rust-lang.org/rustc/targets/custom.html). The exact name of the configuration option is subject to bikeshedding.

Reviewer 6: ```suggestion
Not all compilations targets support embedding this data. Whether the target supports it is recorded in the [target specification JSON](https://doc.rust-lang.org/rustc/targets/custom.html). The exact name of the configuration option is subject to bikeshedding.
```

- Feature Name: `cargo_embed_dependency_versions`
- Start Date: 2019-11-03
- RFC PR: [rust-lang/rfcs#2801](https://github.com/rust-lang/rfcs/pull/2801)
- Rust Issue: None

# Summary
[summary]: #summary

Embed the crate dependency tree in a machine-readable format into compiled binaries so it could be programmatically recovered later.

# Motivation
[motivation]: #motivation

Rust is very promising for security-critical applications due to its safety guarantees, but there currently are gaps in the ecosystem that prevent it. One of them is the lack of any infrastructure for security updates.

> Linux distributions alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.

Reviewer 5: ```suggestion
Linux distributions can alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.
```
My Linux distribution doesn't alert me, alas.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or normal), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.

This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.

> WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.

Reviewer 8: Typo?
```suggestion
WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that records the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
```

Reviewer 9: I believe that's valid English. It's the tooling that makes a record of the hashes.

Reviewer 5: The grammatical number of "record(s)" here depends on that of "tooling", which is singular, entailing "records".  Alternatively, "tooling" could be changed to "tools" to match "record".

Reviewer 10: > tooling that records the hash

In this phrase, I believe "tooling" is a noun and "records" is a verb (present tense), so it's valid the way it is

- Any binary authentication that might be deployed automatically applies to the version information. There is no need to separately authenticate it.
- Tooling for extracting information from binaries (such as ELF sections) is already readily available, as are zlib decompressors and JSON parsers. It can be extracted and parsed [in 5 lines of Python](https://github.com/rust-secure-code/cargo-auditable/blob/master/PARSING.md), or even with a shell one-liner in a pinch.
- This enables third parties such as cloud providers to scan your binaries for you. Google Cloud [already provides such a service](https://cloud.google.com/container-registry/docs/get-image-vulnerabilities), Amazon has [an open-source project you can deploy](https://aws.amazon.com/blogs/publicsector/detect-vulnerabilities-in-the-docker-images-in-your-applications/) while Azure [integrates several partner solutions](https://docs.microsoft.com/en-us/azure/security-center/security-center-vulnerability-assessment-recommendations). They do not support this specific format yet, but integration into Trivy was very easy, so adding support will likely be trivial.

## Why enable this by default?

> If you have 10 programs on your computer that use a library exploitable over the network, your computer and all your data are vulnerable as long as even **a single program** remains vulnerable.

Reviewer 5: ```suggestion
If you have 10 programs on your computer that depend on a library exploitable over the network and use that library in a way that exposes it to exploitation, your computer and your data are vulnerable as long as even **a single program** remains vulnerable.
```
If you'll forgive my pedantry, the buggy library might not be exposed to attackers' input, and some of the data on the computer might be unreachable from the

Reviewer 10: It's out of scope for this RFC, but it'd be nice if Rust tooling could ignore vulnerabilities that are not in the call-graph, the same way the Go tooling does: https://go.dev/blog/vuln

Author: There is ongoing work to implement this in https://google.github.io/osv-scanner/, see https://github.com/google/osv-scanner/pull/452

https://github.com/rustfoundation/painter might also come in handy

## Alternatives

- Do nothing.
  - Identifying vulnerable binaries will remain impossible. We will see increasing number of known vulnerabilities unpatched in production.
- Track version information separately from the binaries, recording it when running `cargo install` and surfacing it through some other Cargo subcommand. When installing not though `cargo install`, rely on Linux package managers to track version information.
  - Identifying vulnerable binaries will remain impossible on all other platforms, as well as on Linux for code compiled with `cargo build`.
  - Verification by third parties will remain impossible.
- Record version information in a `&'static str` in the binary instead if ELF sections, with start/stop markers to allow black-box extraction from the outside.
  - This has been [prototyped](https://github.com/Shnatsel/rust-audit). It has the upside of allowing the binary itself to introspect its version info with little parsing, but the extraction is less efficient, and this is harder to implement and maintain.

> - Record version information in an industry standard SBOM format instead of a custom format.

Reviewer: imo this deserves its own section.  What are the formats and how do they differ?  This doesn't just help people understand this (this is currently relying on "trust me" rather than "come join me in understanding") but it can help reviewers better analyze the proposed format and provide suggestions

Reviewer: e.g this could just be in Prior Art and this could link out to it

>   - This has been prototyped, and we've found the existing formats unsuitable. The primary reasons are a significant binary size increase (the existing formats are quite verbose, not designed for this use case) and issues with reproducible builds (they require timestamps).

Reviewer 11: Just as an FYI that CycloneDX now has a way to do reproducible builds.
https://github.com/CycloneDX/cyclonedx-property-taxonomy/pull/70
The other argument is still valid.

Author: Thanks for the pointer!

I got in touch with some people involved with CycloneDX and I'm investigating if it's possible to select just the right combination of optional fields to have it not blow up the size. 

It is noticeably bigger than the custom format when uncompressed, but gzip does a really good job compacting it. I will revise the RFC with a CycloneDX-compatible format if the prototype shows promise.

Reviewer 11: I'm active in trying to get the https://github.com/CycloneDX/cyclonedx-rust-cargo project going again which is supposed to create CycloneDX SBOMs for Rust projects.
It turns out to be harder than expected so I'm looking at how auditable does it but that doesn't embed all information needed and not all information can be easily recovered from the current JSON format either. I'm torn on how to continue.

But that is a bit separate from this discussion and I don't mean to derail it.

To mitigate the issue you simply need to run the regular security update command provided by your distribution and reboot. The single copy of the library is replaced and the system is secured.

Sadly this does not apply to packages installed from outside the distribution, or software that is not packaged at all.

### Ruby

Ruby code typically has a `Gemfile.lock` which is analogous to `Cargo.lock`. Crucially, Ruby is never compiled into a binary, and this file is present at runtime in the deployed code. So you can tell exactly what versions of which libraries the code you run actually uses.

Thanks to `Gemfile.lock` the code can be automatically checked for vulnerable package versions against databases such as [RubySec](https://rubysec.com/).

Unlike in the Linux distribution scenario, the dependencies of every program have to be updated individually.

### Go

Go statically links all its code, so the C approach of updating a single instance of a shared library is impossible. Every single Go binary contains a copy of the vulnerable code, and all of them have to be rebuilt.

> To make this problem tractable, Go embeds detailed information about the dependency tree in each compiled binary. It is similar to `Cargo.lock` but much more detailed, also including the licenses and other metadata about the dependencies.

Reviewer: Could you get into more detail or link out to it?
- What do they call this section?
- Whats its layout?
- Is there any user control over it and what does it look like?

Much like SBOMs, having more detail on this could help in figuring in bringing reviewers up to a common level of understanding and might spur additional ideas.

Author: Good question! Off the top of my head, https://utcc.utoronto.ca/~cks/space/blog/programming/GoVersionOfYourSource linked up the thread, but I'll try to find more detailed docs on it.

Reviewer 7: It is `.go.buildinfo`. https://pkg.go.dev/runtime/debug#BuildInfo is the full format. Basically the go version, build settings (which includes the checked out git commit) as well as for every module the path, version and checksum.

The dependency info intentionally can't be omitted 
: https://github.com/golang/go/issues/50501#issuecomment-1010225207 Only the git commit seems to be omitable with a flag.

Reviewer: Thanks!

> Replace *[Module](https://pkg.go.dev/runtime/debug#Module) // replaced by this module

Do we do anything about these?  I assume `cargo metadata` (what `cargo auditble` uses) doesn't say anything about these.  I've not used them enough to see what gets included in the `Cargo.lock`

Reviewer: Sounds like go also has native support for displaying this embedded information and lists that as one of the rationals for making it unconfigurable.

> To make this problem tractable, Go embeds detailed information about the dependency tree in each compiled binary. It is similar to `Cargo.lock` but much more detailed, also including the licenses and other metadata about the dependencies.

Reviewer: This proposal is focused on security but I imagine corporate legal audits would also appreciate license information.  Should we consider that to any degree?

The number of times we had random binaries when doing legal audits...

Reviewer 7: You can get the license information by looking the crate versions up on crates.io, right? The tool which extracts the dependency info from the executable can have an option to lookup the license too.

Reviewer: So long as the dependency is from crates.io.

Author: For crates.io, yes. Not all packages originate from crates.io.

The primary use case for this information is cross-referencing versions of the dependencies against [RustSec advisory database](https://github.com/RustSec/advisory-db) and/or [Common Vulnerabilities and Exposures](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). This also enables use cases such as ensuring a fix in a depencency has been propagated across the entirety of your fleet or preventing binaries with unvetted dependencies from accidentally reaching a production environment - all with zero bookkeeping.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Every time an executable is compiled with Cargo, the contents of `Cargo.lock` are embedded in the generated binary. It can be recovered using existing tools like `readelf` or Rust-specific tooling, and then inspected manually or processed in an automated way just like the regular `Cargo.lock` file.

> WASM, asm.js and embedded platforms excempt from this mechanism since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.

Reviewer 12: ```suggestion
WASM, asm.js and embedded platforms are exempt from this mechanism since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
```

  - Identifying vulnerable binaries will remain impossible on all other platforms, as well as on Linux for code compiled with `cargo build`.
  - Verification by third parties will remain impossible.
- Record version information in a `&'static str` in the binary instead if ELF sections, with start/stop markers to allow black-box extraction from the outside.
  - This has been [prototyped](https://github.com/Shnatsel/rust-audit). It has the upside of allowing the binary itself to introspect its version info, but appears to be harder to implement and maintain.
- Provide a Cargo wrapper or plugin to implement this, but do not put it in Cargo itself.
  - When people actually need this information (e.g. to check if they're impacted by a vulnerability) it is too late to reach for third-party tooling - the executables have already been built and deployed, and the information is already lost. As such, this mechanism is completely ineffective if it's not enabled by default.

# Prior art
[prior-art]: #prior-art

The Rust compiler already embeds compiler and LLVM version in the executables built with it. You can see it by running `strings your_executable | grep 'rustc version'`.

> The author is not aware of direct prior art in other languages. Since build system and package management system are usually decoupled, most languages did not have the opportunity to implement anything like this.

Reviewer 12: Go embeds `go.mod` dependency information into its compiled binaries.

I haven't used this for tracking vulnerability info (because Go does not have a rustsec equivalent yet), but it is used for doing license audits: https://github.com/mitchellh/golicense

- A malicious actor could lie about the version information. However, doing so requires modifying the binary - and if a malicious actor can do _that,_ you are pwned anyway. So this does not create any additional attack vectors other than exploiting the tool that's recovering the version information, which can be easily sandboxed.
- Any binary authentication that might be deployed automatically applies to the version information. There is no need to separately authenticate it.
- Tooling for extracting information from binaries (such as ELF sections) is already readily available. Tooling for parsing `Cargo.lock` also exists.
- This enables third parties such as cloud providers to scan your binaries for you. Google Cloud [already provides such a service](https://cloud.google.com/container-registry/docs/get-image-vulnerabilities), Amazon has [an open-source project you can deploy](https://aws.amazon.com/blogs/publicsector/detect-vulnerabilities-in-the-docker-images-in-your-applications/) while Azure [integrates several partner solutions](https://docs.microsoft.com/en-us/azure/security-center/security-center-vulnerability-assessment-recommendations).

> Alternatives:

Reviewer 13: Do we actually need Cargo support for this? Would it be possible to have an empty crate with a `build.rs` script that performs the appropriate analysis and emits the appropriate linker flags? That would certainly defeat the goal of being on-by-default, but it could be a useful first step before Cargo support is ratified and deployed.

Author: AFAIK this can't be fit into a crate because `build.rs` is invoked *before* the rest of the code is compiled. But this almost certainly can be implemented as a Cargo wrapper.

However, it's not very useful unless it's ubiquitous, which is why I really want to see it enabled by default in Cargo itself.

Reviewer 13: When running `build.rs`, Cargo provides the [`CARGO_MANIFEST_DIR`](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts) environment variable. I believe that you could use this to locate the `Cargo.lock` file, and use its location to emit the appropriate linker flags.

Author: I have zero knowledge about using linkers directly (or configuring them), but I would welcome any attempts at prototyping this.

Reviewer 13: Posting this to This Week in Rust's CFP :)

Reviewer 14: I filed https://stackoverflow.com/questions/75052086/how-to-locate-the-cargo-lock-from-build-rs/75052128 about locating Cargo.lock from build.rs . Would love a simpler solution than the approach I have, esp given https://github.com/neilwashere/rust-project-root/pull/18 would break it. `cargo-lock` crate doesnt seem to have a file locator, but I guess something in https://github.com/rustsec/rustsec/ needs to be able to do this.

Author: A full-fledged out-of-tree implementation of this RFC is now available: https://github.com/rust-secure-code/cargo-auditable

- This enables third parties such as cloud providers to scan your binaries for you. Google Cloud [already provides such a service](https://cloud.google.com/container-registry/docs/get-image-vulnerabilities), Amazon has [an open-source project you can deploy](https://aws.amazon.com/blogs/publicsector/detect-vulnerabilities-in-the-docker-images-in-your-applications/) while Azure [integrates several partner solutions](https://docs.microsoft.com/en-us/azure/security-center/security-center-vulnerability-assessment-recommendations). They do not support this specific format yet, but integration into Trivy was very easy, so adding support will likely be trivial.

## Why enable this by default?

If you have 10 programs on your computer that use a library exploitable over the network, your computer and all your data are vulnerable as long as even **a single program** remains vulnerable.

It does not matter if you have discovered and patched 9 out of 10. As long as a single unpatched binary is running, the system is vulnerable.

An attacker needs to only find a single vulnerable program; defense is all-or-nothing. **Every single** vulnerable binary needs to be found and fixed.

> If the mechanism proposed in this RFC is not enabled by default, it will not let you discover all vulnerable binaries in the system, and at that point it might as well not exist.

Reviewer 6: > If the mechanism proposed in this RFC is not enabled by default, it will not let you discover all vulnerable binaries in the system, and at that point it might as well not exist.

You may not wish to escalate this to the point of "if you're never going to enable this by default, reject this RFC entirely", or you may find the RFC rejected entirely.

Author: I will sleep on it and perhaps revise it. Thank you for the feedback.

Author: I've rewritten this part, thank you for pointing it out.
